README
================================================================================
Introducing the Solr PHP Extension
================================================================================
The Solr extension is an extremely fast, light-weight, feature-rich library that 
allows PHP developers to communicate effectively with Solr server instances.

There are tools to add documents and make updates to the solr server.

It also contains tools that allows you to build advanced queries to the server
when searching for documents.

There are also special objects that simplifies the sending of name-value requests
to the server. This includes the SolrParams class and all its child classes.

The SolrClient object allows one to communicate with Solr containers that are 
behind proxy servers or that require HTTP Authentication to proceed.

The SolrClient constructor accepts options such as
http authentication username and passwor, proxy server name, port, login and 
passwords etc.

Using an advanced http client like libcurl allows us to leverage the features
available with the library. We can reuse the HTTP connections without having 
to create a separate one for each request.

================================================================================
How to Install
================================================================================

Please refer to the README.INSTALLATION file.

================================================================================
Magic Methods and Interfaces Implemented
================================================================================

SolrDocument implements the following interfaces.

(1) ArrayAccess  - to access the fields as array keys using field names.
(2) Iterator	 - to iterate over the document fields using foreach()
(3) Serializable - provides custom serialization of the object.

SolrDocument also contains the __get and __set magic methods which allows developers
to access the fields directly. When setting fields, if the field already
has a value the new value will be appended to the list of values for that field.

Each field is a SolrDocumentField object with the following public properties :

(a) name - a string with name of the field.
(b) boost - a double representing the boost value of the field (intentionally empty)
(c) values - an array of all the field values as strings.

Custom Serialization

string SolrDocument::serialize(void) returns an XML document representing a 
SolrDocument as shown below :

<?xml version="1.0" encoding="UTF-8"?>
<solr_document>
  <fields>
    <field name="id">
      <field_value>A0F43D</field_value>
    </field>
    <field name="name">
      <field_value>Israel Ekpo</field_value>
    </field>
    <field name="email">
      <field_value>Israel.Ekpo@israel.ekpo.com</field_value>
    </field>
    <field name="skills">
      <field_value>Reading</field_value>
      <field_value>Writing</field_value>
      <field_value>Soccer</field_value>
      <field_value>Teaching</field_value>
    </field>
    <field name="languages">
      <field_value>Inglés</field_value>
      <field_value>Espanól</field_value>
    </field>
  </fields>
</solr_document>

Here is a complete example of a serialized SolrDocument object:

C:12:"SolrDocument":679:{<?xml version="1.0" encoding="UTF-8"?>
<solr_document>
  <fields>
    <field name="id">
      <field_value>A0F43D</field_value>
    </field>
    <field name="name">
      <field_value>Israel Ekpo</field_value>
    </field>
    <field name="email">
      <field_value>Israel.Ekpo@israel.ekpo.com</field_value>
    </field>
    <field name="skills">
      <field_value>Reading</field_value>
      <field_value>Writing</field_value>
      <field_value>Soccer</field_value>
      <field_value>Teaching</field_value>
    </field>
    <field name="languages">
      <field_value>Inglés</field_value>
      <field_value>Espanól</field_value>
    </field>
  </fields>
</solr_document>
}

One of the items on my todo list is to create a response writer to return 
serialized SolrDocument objects instead of document arrays.

SolrDocument::unserialize(string $serialized) accepts an XML document 
representing the SolrDocument object. It will the bring the object back to live.

The SolrDocument class also has the method SolrDocument::getInputDocument() to 
allow one do get the SolrInputDocument version of a SolrDocument instance.

This method may be helpful if one needs to resubmit the document to the server to 
updates.

The Solr extension has the SolrQuery object (a child of SolrParams) that enables
the developer to send custom advanced name-value requests to the solr server.

The SolrQuery object can also be serialized and reused later, which makes it very
helpful to saving the state of the application across multiple requests. This may be
very useful in cases such as facet browsing where additional parameters may need to be
added to the current object or removed from it to get the desired results without
having to start over from scratch.

================================================================================
Parsing of XML Responses from the Solr Server
================================================================================
XML responses from the solr server are expected to be formatted using version 2.2
These xml documents are parsed into serialized php code and returned as
read-only SolrObject instances whose properties can also be accessed as array
keys in addition to being accessible directly via the object->member notation.

Having the properties accessible via object[member] notation is helpful in cases
where the property name is not valid (contains dots and other characters not 
legal in php.

================================================================================
How to Report Bugs
================================================================================

Please report bugs to iekpo@php.net

If you experience a crash due to a segmentation fault, please follow the instructions on the link below 
to get a gdb backtrace and then submit the trace in your bug report as well

http://bugs.php.net/bugs-generating-backtrace.php

Thank you for using PHP


CHANGELOG
1.0.3 : (In progress)

- Applied patch to handle SOLR_DEBUG better
- No longer pass "waitFlush" in commit
- Organized versioning in the .h files
- Ditched "phpnative" responseWriter, introduced "phps" ("php" native still in progress)

1.0.2 : 2011-11-29

- Several Bug Fixes by Pierrick, Felipe and Tony
- Changes to make the code compile on 5.2.3 and greater (iekpo)
- Changes to Tests by (pierrick)

1.0.1 : 2011-06-04

- Added support for json response writer in SolrClient
- Removed final bit from classes so that they can be mocked in unit tests
- Changed from beta to stable
- Included phpdoc stubs in source to enable autocomplete of Solr classes and methods in IDE during development
- Lowered libxml2 version requirement to 2.6.16 

0.9.11 : 2010-06-22

- Added ability to specify response writer in constructor option
- Added new method to set response writer SolrClient::setResponseWriter() 
- Currently, the only supported response writers are 'xml' and 'phpnative'
- Added support for new native Solr response writer
- New response writer is available at https://issues.apache.org/jira/browse/SOLR-1967

0.9.10 : 2010-05-04

- Increased compatibility with older systems running CentOS 4 or 5 and RHEL4 or 5
- Added ability to compile directly without having to build libcurl and libxml2 from source	on older systems			
- Lowered minimum supported version for libcurl to 7.15.0 (Alex Samorukov)
- Lowered minimum supported version for libxml2 to 2.6.26 (Alex Samorukov)
- Fixed PECL Bug# 17172 MoreLikeThis only parses one doc (trevor at blubolt dot com, max at blubolt dot com)
- Declared workaround macros for SSL private key constants due to support for earlier versions of libcurl (Alex Samorukov)
- Changed extension version numbers to start using hexadecimal numbers (Israel Ekpo)
- Added instructions on how to attempt to compile on windows (Israel Ekpo)
- Fixed PECL Bug# 17292 sending UTF-8 encoding in header (giguet at info dot unicaen dot fr)

0.9.9 : 2010-01-10

- Fixed Bug #17009 Creating two SolrQuery objects leads to wrong query value
- Reset the buffer for the request data from the previous request in SolrClient
- Added new internal static function solr_set_initial_curl_handle_options()
- Moved the intialization of CURL handle options to solr_set_initial_curl_handle_options() function
- Resetting the CURL options on the (CURL *) handle after each request is completed
- Added more explicit error message to indicate that cloning SolrParams objects and its descendants is currently not yet supported

0.9.8 : 2009-12-04

- Fixed config.w32 for Windows build support (Pierre, Pierrick)
- Fixed Bug #16943 Segmentation Fault from solr_encode_string() during attempt to retrieve solrXmlNode->children->content when solrXmlNode->children is NULL (Israel)
- Disabled Expect header in libcurl (Israel)
- Disabled Memory Debugging when normal debug is enabled (Israel)
- Added list of contributors to the project

0.9.7 : 2009-11-17

- Fixed bug 16924 AC_MSG_NOTICE() is undefined in autoconf 2.13
- Added new method SolrClient::getDebug()
- Modified SolrClient::__construct() so that port numbers and other integer values for the options can be passed as strings.
- Changed internal string handling mechanism to allow for tracking of memory allocation in debug mode.
- Lowered minimum php version to 5.2.3. Unfortunately, this is the lowest PHP version that will be supported. PHP versions lower than 5.2.3 are not compatible or are causing tests to FAIL.
- Added php stubs for code-completion assists in IDEs and editors.
- Added more examples

0.9.6 : 2009-11-01

- Reduced minimum php version to 5.2.4

0.9.5 : 2009-10-27

- Added support for secure connections to SSL-enabled Solr servers (SSL/HTTPS). 
- Made connection errors more verbose.
- The following options where added to the SolrClient constructor :
 - secure          (Boolean value indicating whether or not to connect in secure mode)
 - timeout         (This is maximum time in seconds allowed for the http data transfer operation. Default is 30 seconds)
 - ssl_cert        (File name to a PEM-formatted file containing the private key + private certificate (concatenated in that order) )
 - ssl_key         (File name to a PEM-formatted private key file only)
 - ssl_keypassword (Password for private key)
 - ssl_cainfo      (Name of file holding one or more CA certificates to verify peer with)
 - ssl_capath      (Name of directory holding multiple CA certificates to verify peer with )	

0.9.4 : 2009-10-23
 - Fixed bug 16897 - ap_php_vasprintf not available in 5.2.11

0.9.3 : 2009-10-19

- Corrected configuration options for automated installation via PECL. Removed extra hyphens.

0.9.2 : 2009-10-19
				
- Fixed bug #16874 Added explanation in documentation
- Updated SolrParams::__toString() - modified to return name-value pair list instead of serialized object string
- Added the following methods to SolrParams :
    SolrParams::toString()
    SolrParams::getParam()
- Created the following aliases in SolrParams : 
    SolrParams::set()
    SolrParams::add()
    SolrParams::get()
- Added the following methods to SolrQuery :
    SolrQuery::getQuery()
    SolrQuery::getStart()
    SolrQuery::getRows()
    SolrQuery::getFields()
    SolrQuery::getSortFields()
    SolrQuery::getFilterQueries()
    SolrQuery::getTimeAllowed()
    SolrQuery::getFacet()
    SolrQuery::getFacetFields()
    SolrQuery::getFacetQueries()
    SolrQuery::getFacetPrefix()
    SolrQuery::getFacetSort()
    SolrQuery::getFacetLimit()
    SolrQuery::getFacetOffset()
    SolrQuery::getFacetMinCount()
    SolrQuery::getFacetMissing()
    SolrQuery::getFacetMethod()
    SolrQuery::getFacetDateFields()
    SolrQuery::getFacetDateStart()
    SolrQuery::getFacetDateEnd()
    SolrQuery::getFacetDateGap()
    SolrQuery::getFacetDateHardEnd()
    SolrQuery::getFacetDateOther()
    SolrQuery::getHighlight()
    SolrQuery::getHighlightFields()
    SolrQuery::getHighlightSnippets()

solr_constants.h
/*
   +----------------------------------------------------------------------+
   | PHP Version 5                                                        |
   +----------------------------------------------------------------------+
   | Copyright (c) 1997-2009 The PHP Group                                |
   +----------------------------------------------------------------------+
   | This source file is subject to version 3.01 of the PHP license,      |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | http://www.php.net/license/3_01.txt                                  |
   | If you did not receive a copy of the PHP license and are unable to   |
   | obtain it through the world-wide-web, please send a note to          |
   | license@php.net so we can mail you a copy immediately.               |
   +----------------------------------------------------------------------+
   | Author: Israel Ekpo <iekpo@php.net>                                  |
   +----------------------------------------------------------------------+
*/

/* $Id: solr_constants.h 311799 2011-06-04 08:05:00Z iekpo $ */

#ifndef SOLR_CONSTANTS_H
#define SOLR_CONSTANTS_H

/* {{{ Current version of Solr extension */
#define PHP_SOLR_DOTTED_VERSION "1.0.3-alpha"
#define SOLR_CLIENT_USER_AGENT "PHP Solr Client " PHP_SOLR_DOTTED_VERSION

#define SOLR_SPACE_STRING " "

/* {{{ Constants for the SolrClient with*/
#define SOLR_DEFAULT_UPDATE_SERVLET  "update"
#define SOLR_DEFAULT_SEARCH_SERVLET  "select"
#define SOLR_DEFAULT_THREADS_SERVLET "admin/threads"
#define SOLR_DEFAULT_PING_SERVLET    "admin/ping"
#define SOLR_DEFAULT_TERMS_SERVLET   "terms"

#define SOLR_PHP_SERIALIZED_RESPONSE_WRITER    "phps"
#define SOLR_XML_RESPONSE_WRITER               "xml"
#define SOLR_JSON_RESPONSE_WRITER              "json"

#define SOLR_REQUEST_DEFAULT_HOST "localhost"
#define SOLR_REQUEST_DEFAULT_PATH "solr"
#define SOLR_REQUEST_DEFAULT_QS_DELIMITER "&"
#define SOLR_REQUEST_DEFAULT_PORT 8983L

#define SOLR_CURL_MAX_WRITE_SIZE CURL_MAX_WRITE_SIZE
#define SOLR_CURL_OPTIONS_LENGTH 512
/* }}} */

/* {{{ Names of Objects */
#define PHP_SOLR_OBJECT_CLASSNAME      	   	   "SolrObject"
#define PHP_SOLR_DOCUMENT_CLASSNAME      	   "SolrDocument"
#define PHP_SOLR_DOCUMENT_FIELD_CLASSNAME      "SolrDocumentField"
#define PHP_SOLR_INPUT_DOCUMENT_CLASSNAME      "SolrInputDocument"
#define PHP_SOLR_CLIENT_CLASSNAME              "SolrClient"
#define PHP_SOLR_PARAMS_CLASSNAME              "SolrParams"
#define PHP_SOLR_QUERY_CLASSNAME               "SolrQuery"
#define PHP_SOLR_MODIFIABLE_PARAMS_CLASSNAME   "SolrModifiableParams"
#define PHP_SOLR_RESPONSE_CLASSNAME   		   "SolrResponse"
#define PHP_SOLR_QUERY_RESPONSE_CLASSNAME      "SolrQueryResponse"
#define PHP_SOLR_UPDATE_RESPONSE_CLASSNAME     "SolrUpdateResponse"
#define PHP_SOLR_PING_RESPONSE_CLASSNAME   	   "SolrPingResponse"
#define PHP_SOLR_GENERIC_RESPONSE_CLASSNAME    "SolrGenericResponse"
#define PHP_SOLR_UTILS_CLASSNAME           	   "SolrUtils"

#define PHP_SOLR_EXCEPTION_CLASSNAME           "SolrException"
#define PHP_SOLR_ILLEGAL_OPERATION_EXCEPTION_CLASSNAME	"SolrIllegalOperationException"
#define PHP_SOLR_ILLEGAL_ARGUMENT_EXCEPTION_CLASSNAME	"SolrIllegalArgumentException"
#define PHP_SOLR_CLIENT_EXCEPTION_CLASSNAME     "SolrClientException"

/* }}} */

/* {{{ Informational constants */
#define PHP_SOLR_EXTENSION_NAME   "solr"
#define PHP_LIBXML_EXTENSION_NAME "libxml"

/* }}} */

/* {{{ Property names */
#define SOLR_INDEX_PROPERTY_NAME 		"_hashtable_index"
#define SOLR_FIELD_NAME_PROPERTY_NAME   "name"
#define SOLR_FIELD_BOOST_PROPERTY_NAME  "boost"
#define SOLR_FIELD_VALUES_PROPERTY_NAME "values"

#define SOLR_SOURCELINE_NO_PROPERTY_NAME "sourceline"
#define SOLR_SOURCEFILE_PROPERTY_NAME "sourcefile"
#define SOLR_ZIFNAME_PROPERTY_NAME "zif_name"

#ifdef PHP_WIN32
#define __func__ __FUNCTION__
#endif

#define SOLR_FILE_LINE_FUNC __FILE__, __LINE__, __func__
/* }}} */

/* {{{ Constants for Method/Function arguments and return styles */
#define SOLR_ARG_PASS_BY_REF_FALSE 0
#define SOLR_ARG_PASS_BY_REF_TRUE  1

#define SOLR_ARG_PASS_REMAINING_BY_REF_FALSE 0
#define SOLR_ARG_PASS_REMAINING_BY_REF_TRUE  1

#define SOLR_ARG_ALLOW_NULL_FALSE 0
#define SOLR_ARG_ALLOW_NULL_TRUE  1

#define SOLR_METHOD_RETURN_REFERENCE_FALSE 0
#define SOLR_METHOD_RETURN_REFERENCE_TRUE  1
/* }}} */

/* {{{ Constants for Sorting SolrDocument Fields */
#define SOLR_SORT_ASC  1
#define SOLR_SORT_DESC 2

#define SOLR_SORT_FIELD_NAME 1
#define SOLR_SORT_FIELD_VALUE_COUNT 2
#define SOLR_SORT_FIELD_BOOST_VALUE 4
/* }}} */

/* {{{ Constants for document encoding formats */
#define SOLR_XML_ENCODE_OBJ 0L
#define SOLR_XML_ENCODE_DOC 1L
/* }}} */

/**
 * Setting this to 8 gets rid of the extra while loops in _zend_hash_init()
 *
 * _zend_hash_init loops 3 times to increase the value of the size to 8
 */
#define SOLR_INITIAL_HASH_TABLE_SIZE (1<<3)

/* {{{ Memory Allocation Related Constants */
#define SOLR_HASHTABLE_PERSISTENT 0
#define SOLR_DOCUMENT_PERSISTENT 0
#define SOLR_DOCUMENT_FIELD_PERSISTENT 0
#define SOLR_CLIENT_PERSISTENT 0
#define SOLR_PARAMS_PERSISTENT 0
/* }}} */

/* {{{ SolrException Codes : Internal and UserSpace */

/* {{{ Error codes and messages */

#define SOLR_ERROR_1000 1000L
#define SOLR_ERROR_1000_MSG "Error un-serializing response"

#define SOLR_ERROR_1001 1001L
#define SOLR_ERROR_1001_MSG "SolrClient objects cannot be serialized or unserialized"

#define SOLR_ERROR_1002 1002L
#define SOLR_ERROR_1002_MSG "The '%s' property cannot be removed or set to NULL. SolrObject properties cannot be unset or set to NULL."

#define SOLR_ERROR_1003 1003L
#define SOLR_ERROR_1003_MSG "Error building HTTP query from parameters"

#define SOLR_ERROR_1004 1004L
#define SOLR_ERROR_1004_MSG "Unsuccessful HTTP request"

#define SOLR_ERROR_1005 1005L
#define SOLR_ERROR_1005_MSG "SolrInputDocument objects cannot be serialized or unserialized"

#define SOLR_ERROR_1006 1006L
#define SOLR_ERROR_1006_MSG "SolrObject instances are read-only. Properties cannot be updated or removed."

#define SOLR_ERROR_1007 1007L
#define SOLR_ERROR_1007_MSG "SolrDocumentField instances are read-only. Properties cannot be updated or removed."

#define SOLR_ERROR_1008 1008L
#define SOLR_ERROR_1008_MSG "Internal Error 1008 generated from %s %d %s. The observed error is a possible side-effect of an illegal/unsupported operation in userspace. Please check the documentation and try again later."

#define SOLR_ERROR_1009 1009L
#define SOLR_ERROR_1009_MSG "The observed error is a possible side-effect of an illegal/unsupported operation in userspace. This object cannot be cloned. Cloning of this object type is not yet supported."


#define SOLR_ERROR_4000 4000L
#define SOLR_ERROR_4000_MSG  "Invalid argument"

#define SOLR_ERROR_4001 4001L
#define SOLR_ERROR_4001_MSG  "Illegal Operation"

#define SOLR_ERROR_4002 4002L
#define SOLR_ERROR_4002_MSG  "Return value requested without processing output"

/* }}} */

/* {{{ UserSpace Errors */

/* }}} */

/* {{{ Compile time errors */

/* }}} */

/* }}} */

#endif /* SOLR_CONSTANTS_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: fdm=marker
 * vim: noet sw=4 ts=4
 */
